### DISCUSSION TITLE
The second thing that I miss in Odin is Operator Overloading. And that make me confused.

### DISCUSSION THREAD

**Discussion by joaocarvalhoopen** [odin-lang/Odin#2731] [Q&A]
Hello,

I really like the design of Odin, and it's design decision, it's simplicity and elegance, but I must not be seeing something in Odin, because in my brain is confused, it's not making sense that Odin is used for high performance computing and that gingerBill is a Physicist and uses it for high performance mathematical computation of real time fire, smoke and water special effect simulations in the CPU and GPU and Odin doesn't have Operator Overloading for any user structure of custom math ?!??!??!!?

I'm I missing something?

This two realities seem worlds apart, so I must be missing or not understanding something?

Can please someone help to clarify my mind?

Thank you very much.

The very best regards,
João Carvalho

**Comment by drmargarido** [comment]
There is no operator overloading, but there is support for array programming with vector and matrix math operations on the base language itself. [Link](https://odin-lang.org/docs/overview/#array-programming)

**Comment by joaocarvalhoopen** [comment]
Hello, thank you drmargarido for the fast reply,

I have read the array programming section and the Matrix section of seen the matrix and array operation overloading, but the is steel one aspect that is not compatible of using Odin in language operator overloading and that is this citation that I put here from the manual:

"
Currently, matrices are limited to a maximum of 16 elements (rows*columns), and a minimum of 1 element. This is because matrices are stored as values (not a reference type), and thus operations on them will be stored on the stack. Restricting the maximum element count minimizes the possibility of stack overflows.
"

There should be support for unlimited size of matrix's of any type of data, ex: f32, f64, f128, complex32 and complex64, complex 128.


I'm thinking about scientific data calculations or processing and not only computer graphical matrix transformations.

Thank you very much,

Best regards,
João Carvalho

**Comment by laytan** [comment]
From the FAQ:

Why does Odin not have operator overloading? [#](https://odin-lang.org/docs/faq/#why-does-odin-not-have-operator-overloading)

The design goals of Odin were explicitness and simplicity. Operator overloading is very easily abused and can be used to do many magical things. A procedure is clearer and more explicit.

Array programming is available in Odin; this removes some of the need for operator overloading when creating mathematical libraries.

**Comment by Kelimion** [comment]
> There should be support for unlimited size of matrix's of any type of data, ex: f32, f64, f128, complex32 and complex64, complex 128.

Why? The number one scientific language (Python) doesn't have it without resorting to a third-party package. On what theory then should Odin have it as a base type?

**Comment by joaocarvalhoopen** [comment]
Hello Kelimion,

thank you for the rapid response.
I don't think that Odin should have it as a Base Type, the thing is that if it is not a base type it can't have operation overloading like NumPy has in Python :-)

Note: Like I said, I like almost all decisions in the design of Odin, they are really nice and simple, especially when compared to C++ or Rust, in my list the one that lacks is Operator Overloading for the implementation of a good mathematical library , but I know that it can be improperly used and that not having it will simplify enormously the source code of the language.
I'm currently spreading the word of Odin to friends and colleges :-)

Best regards,
João Carvalho

**Comment by David-OConnor** [comment]
I have the same concern as the OP of the issue: The areas where I want a high-performance language overlap heavily with the ones where I want operator overloading!

**Reply by DerTee** [reply]
I'm guessing the issue is simply that not every language's tradeoffs are ideal for everyone who wants to use it. Odin's main strength is its clear focus in my opinion. While I don't like every single thing about Odin, it is the language I currently enjoy the most if I use it for its main focus: High performance game/application programming.

If I do one-off string processing stuff, then Odin is not the best choice. Python is much better suited for that (unless there's a lot of data and a lot of processing). If I do simple backend web dev, then I use PHP, which is purpose built for exactly that (maybe I'll use Golang in the future). If I want to write a website frontend, I use Elm (or JS in a pinch), which is built for exactly that.

Maybe Julia is the right language for you. I've tried it and had a good time, but I would really want Julia to produce standalone executables, but that is not a priority for most scientists, so the devs don't care very much about this feature, which is entirely reasonable.

**Reply by David-OConnor** [reply]
I do a mix of scientific computing, computer graphics, and embedded; all of which make heavy use of vectors, quaternions and matrices. Operator overloading is very nice in those cases. I think Julia is not a serious contender in these domains, outside some scientific uses where you're not building applications.

Odin and Jai sound very promising in these domains! I currently primarily use rust, and would love for it to have competition.

Of note, Jai supports operator overloading, which isn't surprising given it's initial domain.

**Reply by DerTee** [reply]
I don't understand your motivation. Why do you want competition in the space of programming languages for your use cases? To have a harder time choosing which one to use? If you know one language well that is suited to your problem domain, then use that. If Rust works well for you, use it and maybe evaluate the alternatives once a year if you want. If Jai works well for you, then use that.

Why do you need Odin to be the same as the other languages when you can just use the other languages? Sure, there might be some things you like better in Odin, but you will never find the absolutely ideal language, that has only the exact things you want and none of the downsides. Even Bill isn't happy with every single thing in Odin and he wrote it more or less for his own needs.

**Comment by flysand7** [comment]
Regarding matrices, I think it's better for everyone if they're used as part of some third-party package, rather than being built into the compiler itself.

My reasoning stems from the fact that yes, it would be very much possible to use the basic matrix algorithms into the compiler, but if you wanted really large matrices, you don't really want basic algorithms - they're quadratic and slow. And if there's a new cutting-edge matrix factoring algorithm, the compiler would need to be updated. Or have different versions of matrices depending on your use-case. You also have no way of using CPU-bound builtin matrices on a GPU, which I assume would be what you'd be doing for advanced use-cases such as machine learning, that is, if you care about not wasting extra weeks on CPU computations.

The second thing I think I should mention is that the reason matrices are only supported up to the 4x4 size is because these are trivially optimized by the compiler into SIMD code and are common enough in computer graphics (which is the primary domain for Odin, but it's not exclusive to that) to be useful.

**Reply by pdenapo** [reply]
I think that operator overloading would be very useful [and that matrices and quaternions are not a substitute for it].  For instance, odin does not have support for exact decimal numbers [which are very important for instance for finance applications]. Consider for example the library mpdecimal https://www.bytereef.org/mpdecimal/index.html - I have created a wrapper for D using it https://github.com/pdenapo/d_mpdecimal which allow me to write D code using mpdecimal in a natural and legible way. This is just impossible in Odin due to the absence of operator overloading!
